File: apps/mobile/lib/features/geofences/data/models/geofence_assignment_model.dart
import '../../domain/entities/geofence_assignment.dart';

/// JSON mappers for the geofence-assignment API
/// (`apps/api/src/attendance/dto/geofence-assignment.dto.ts`).
class GeofenceRefModel extends GeofenceRef {
  const GeofenceRefModel({required super.id, required super.name});

  factory GeofenceRefModel.fromJson(Map<String, dynamic> json) =>
      GeofenceRefModel(
        id: json['id'] as String? ?? '',
        name: json['name'] as String? ?? '',
      );
}

class AssignedEmployeeModel extends AssignedEmployee {
  const AssignedEmployeeModel({
    required super.id,
    required super.assignmentId,
    required super.employeeCode,
    required super.firstName,
    required super.lastName,
    super.displayName,
    super.assignedBy,
  });

  factory AssignedEmployeeModel.fromJson(Map<String, dynamic> json) =>
      AssignedEmployeeModel(
        id: json['id'] as String? ?? '',
        assignmentId: json['assignmentId'] as String? ?? '',
        employeeCode: json['employeeCode'] as String? ?? '',
        firstName: json['firstName'] as String? ?? '',
        lastName: json['lastName'] as String? ?? '',
        displayName: json['displayName'] as String?,
        assignedBy: json['assignedBy'] as String?,
      );
}

class AssignableEmployeeModel extends AssignableEmployee {
  const AssignableEmployeeModel({
    required super.id,
    required super.employeeCode,
    required super.firstName,
    required super.lastName,
    super.displayName,
    super.currentGeofences,
    super.requiresReassign,
  });

  factory AssignableEmployeeModel.fromJson(Map<String, dynamic> json) =>
      AssignableEmployeeModel(
        id: json['id'] as String? ?? '',
        employeeCode: json['employeeCode'] as String? ?? '',
        firstName: json['firstName'] as String? ?? '',
        lastName: json['lastName'] as String? ?? '',
        displayName: json['displayName'] as String?,
        currentGeofences: ((json['currentGeofences'] as List<dynamic>?) ??
                const [])
            .map((e) => GeofenceRefModel.fromJson(e as Map<String, dynamic>))
            .toList(),
        requiresReassign: json['requiresReassign'] as bool? ?? false,
      );
}

class AssignableEmployeesPageModel extends AssignableEmployeesPage {
  const AssignableEmployeesPageModel({
    required super.employees,
    required super.allowMultipleGeofencesPerEmployee,
    required super.tenantHasEmployees,
    required super.total,
  });

  factory AssignableEmployeesPageModel.fromJson(Map<String, dynamic> json) =>
      AssignableEmployeesPageModel(
        employees: ((json['employees'] as List<dynamic>?) ?? const [])
            .map((e) =>
                AssignableEmployeeModel.fromJson(e as Map<String, dynamic>))
            .toList(),
        allowMultipleGeofencesPerEmployee:
            json['allowMultipleGeofencesPerEmployee'] as bool? ?? false,
        tenantHasEmployees: json['tenantHasEmployees'] as bool? ?? false,
        total: json['total'] as int? ?? 0,
      );
}

class AssignResultModel extends AssignResult {
  const AssignResultModel({
    required super.assigned,
    required super.skipped,
    required super.reassigned,
  });

  factory AssignResultModel.fromJson(Map<String, dynamic> json) =>
      AssignResultModel(
        assigned: json['assigned'] as int? ?? 0,
        skipped: json['skipped'] as int? ?? 0,
        reassigned: json['reassigned'] as int? ?? 0,
      );
}

class UnassignResultModel extends UnassignResult {
  const UnassignResultModel({
    required super.removed,
    required super.leftWithoutGeofence,
  });

  factory UnassignResultModel.fromJson(Map<String, dynamic> json) =>
      UnassignResultModel(
        removed: json['removed'] as int? ?? 0,
        leftWithoutGeofence:
            ((json['leftWithoutGeofence'] as List<dynamic>?) ?? const [])
                .map((e) => e.toString())
                .toList(),
      );
}

class GeofenceCoverageModel extends GeofenceCoverage {
  const GeofenceCoverageModel({
    required super.countsByGeofence,
    required super.totalEmployees,
    required super.unassignedEmployees,
    required super.tenantHasEmployees,
  });

  factory GeofenceCoverageModel.fromJson(Map<String, dynamic> json) {
    final raw = (json['countsByGeofence'] as Map<String, dynamic>?) ?? const {};
    return GeofenceCoverageModel(
      countsByGeofence: raw.map(
        (key, value) => MapEntry(key, (value as num?)?.toInt() ?? 0),
      ),
      totalEmployees: json['totalEmployees'] as int? ?? 0,
      unassignedEmployees: json['unassignedEmployees'] as int? ?? 0,
      tenantHasEmployees: json['tenantHasEmployees'] as bool? ?? false,
    );
  }
}

File: apps/admin/src/app/pages/settings/geofence-assignment.model.ts
/**
 * Employee ↔ geofence assignment contracts (mirrors
 * apps/api/src/attendance/dto/geofence-assignment.dto.ts).
 *
 * Attendance is scoped to the geofences an employee is assigned to; an
 * employee with no assignment cannot punch anywhere, which is why the
 * unassigned counts below are surfaced prominently rather than buried.
 */

export interface GeofenceRef {
  readonly id: string;
  readonly name: string;
}

export interface AssignedEmployee {
  readonly id: string;
  readonly assignmentId: string;
  readonly employeeCode: string;
  readonly firstName: string;
  readonly lastName: string;
  readonly displayName: string | null;
  readonly photograph: string | null;
  readonly employmentStatus: string;
  readonly assignedAt: string;
  /** Null marks a row created by the backfill migration, not by an admin. */
  readonly assignedBy: string | null;
}

export interface AssignedEmployeesResponse {
  readonly geofence: GeofenceRef;
  readonly total: number;
  readonly employees: readonly AssignedEmployee[];
}

export interface AssignableEmployee {
  readonly id: string;
  readonly employeeCode: string;
  readonly firstName: string;
  readonly lastName: string;
  readonly displayName: string | null;
  readonly photograph: string | null;
  /** Geofences other than the one being edited. */
  readonly currentGeofences: readonly GeofenceRef[];
  /** True when adding this employee means moving them off another geofence. */
  readonly requiresReassign: boolean;
}

export interface AssignableEmployeesResponse {
  readonly total: number;
  readonly page: number;
  readonly pageSize: number;
  readonly allowMultipleGeofencesPerEmployee: boolean;
  /**
   * False means the tenant has no employees at all — distinct from a search
   * that simply matched nothing, and the trigger for the "create employees
   * first" empty state.
   */
  readonly tenantHasEmployees: boolean;
  readonly employees: readonly AssignableEmployee[];
}

export interface AssignConflict {
  readonly employeeId: string;
  readonly employeeName: string | null;
  readonly currentGeofenceId: string;
  readonly currentGeofenceName: string;
}

export interface AssignResult {
  readonly geofenceId: string;
  readonly assigned: number;
  readonly skipped: number;
  readonly reassigned: number;
}

export interface UnassignResult {
  readonly geofenceId: string;
  readonly removed: number;
  /** These employees can no longer punch anywhere until reassigned. */
  readonly leftWithoutGeofence: readonly string[];
}

export interface GeofenceCoverage {
  readonly countsByGeofence: Readonly<Record<string, number>>;
  readonly totalEmployees: number;
  readonly unassignedEmployees: number;
  readonly tenantHasEmployees: boolean;
}

export interface GeofenceAssignmentPolicy {
  readonly allowMultipleGeofencesPerEmployee: boolean;
  /** Present on update: employees holding more than one assignment already. */
  readonly employeesOverLimit?: number;
}
